packages/@aws-cdk/aws-lambda-nodejs/lib/bundling.ts

```diff
@@ -150,69 +166,179 @@ export class Bundling {
-      fs.writeFileSync(
-        path.join(projectRoot, '.package.json'),
-        JSON.stringify({ dependencies }),
-      );
-
-      // Use npm unless we have a yarn.lock.
-      if (fs.existsSync(path.join(projectRoot, LockFile.YARN))) {
-        installer = Installer.YARN;
-        lockfile = LockFile.YARN;
-      } else if (fs.existsSync(path.join(projectRoot, LockFile.NPM))) {
-        lockfile = LockFile.NPM;
-      }
-
-      // Move dummy package.json and lock file then install
-      depsCommand = chain([
-        `mv ${cdk.AssetStaging.BUNDLING_INPUT_DIR}/.package.json ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}/package.json`,
-        lockfile ? `cp ${cdk.AssetStaging.BUNDLING_INPUT_DIR}/${lockfile} ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}/${lockfile}` : '',
-        `cd ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR} && ${installer} install`,
-      ]);
+    // Local
+    let localBundler: cdk.ILocalBundling | undefined;
+    if (!options.forceDockerBundling) {
+      localBundler = new LocalBundler({
+        projectRoot,
+        relativeEntryPath,
+        parcelOptions: options,
+        dependencies,
+        installer,
+        lockFile,
+      });
     }
 
+    // Docker
+    const dockerBundler = new DockerBundler({
+      relativeEntryPath,
+      parcelOptions: options,
+      dependencies,
+      installer,
+      lockFile,
+    });
+
     return lambda.Code.fromAsset(projectRoot, {
       assetHashType: cdk.AssetHashType.BUNDLE,
       bundling: {
-        image,
-        command: ['bash', '-c', chain([parcelCommand, depsCommand])],
-        environment: options.parcelEnvironment,
-        volumes: options.cacheDir
-          ? [{ containerPath: '/parcel-cache', hostPath: options.cacheDir }]
-          : [],
-        workingDirectory: path.dirname(containerEntryPath).replace(/\\/g, '/'), // Always use POSIX paths in the container
+        local: localBundler,
+        ...dockerBundler.bundlingOptions,
       },
     });
   }
 }
 
+interface BundlerProps {
+  relativeEntryPath: string;
+  parcelOptions: ParcelOptions;
+  dependencies?: { [key: string]: string };
+  installer: Installer;
+  lockFile?: LockFile;
+}
+
+interface LocalBundlerProps extends BundlerProps {
+  projectRoot: string;
+}
+
+/**
+ * Local Parcel bundler
+ */
+class LocalBundler implements cdk.ILocalBundling {
+  public static get runsLocally(): boolean {
+    if (LocalBundler._runsLocally !== undefined) {
+      return LocalBundler._runsLocally;
+    }
+    if (os.platform() === 'win32') { // TODO: add Windows support
+      return false;
+    }
+    try {
+      const parcel = spawnSync(require.resolve('parcel'), ['--version']);
+      LocalBundler._runsLocally = /^2/.test(parcel.stdout.toString().trim()); // Cache result to avoid unnecessary spawns
+      return LocalBundler._runsLocally;
+    } catch {
+      return false;
+    }
+  }
+
+  private static _runsLocally?: boolean;
+
+  constructor(private readonly props: LocalBundlerProps) {}
```

Comment:
Do we really need to pass all the options here? I recall `tryBundle`'s second argument is `options` which should include all the needed context.

To verify we designed the protocol correctly, let's try to implement `tryBundle` through a simple "lambda-interface" instead of as a class:

```ts
local: {
  tryBundle: (outdir: string, options: ...) => {
    // foo bar
  }
}
```
```

Reply:
I will give it a shot but it seemed easier/cleaner to implement it with the bundler's option instead of parsing the Docker command. Might have implications to offer Windows support for local bundling also.

Reply:
Oh I see. Yeah I wouldn't parse the docker command...

You are right. 

Let's leave as is. I am good